### DISCUSSION TITLE
__EMSCRIPTEN_major/minor/tiny__ defined for compile js ?

### DISCUSSION THREAD

**Discussion by Thaina** [emscripten-core/emscripten#19168] [Ideas]
There is an major update about EMVAL related functionality in emscripten in version 2 and 3. And I am in needed to write a js code that can support both version

Something like this

```js
/*global requireHandle, _emval_register, _emval_run_destructors*/

var myLib = {
  JS_GetDate__deps: ['_emval_register'],
  JS_GetDate: function() {
    return __emval_register(new Date());
  },
  
  JS_GetTime__deps: ['$requireHandle'],
  JS_GetTime: function(/** @type {Date} */date) {
    date = requireHandle(date);
    return date.getTime();
  },
  
  JS_Free__deps: ['_emval_run_destructors'],
  JS_Free: function(/** @type {Date} */date) {
    __emval_run_destructors(date);
  },
}

mergeInto(LibraryManager.library,myLib);
```

I wish I could writing a code like `#if EMSCRIPTEN_VERSION < 2.2.0` to switch between code that depend and use `_emval_register` and code that depend and use use `Emval` object when anyone bring this code to use in difference emscripten version

```js
/*global requireHandle, _emval_register, _emval_run_destructors*/

var myLib = {
#if EMSCRIPTEN_VERSION < 2.2.0
  JS_GetDate__deps: ['_emval_register'],
#endif
  JS_GetDate: function() {
#if EMSCRIPTEN_VERSION < 2.2.0
    return __emval_register(new Date());
#else
    return Emval.toHandle(new Date());
#endif
  },
  
#if EMSCRIPTEN_VERSION < 2.2.0
  JS_GetTime__deps: ['$requireHandle'],
#endif
  JS_GetTime: function(/** @type {Date} */date) {
#if EMSCRIPTEN_VERSION < 2.2.0
    date = requireHandle(date);
#else
    date = Emval.toValue(date);
#endif
    return date.getTime();
  },
  
  JS_Free__deps: ['_emval_run_destructors'],
  JS_Free: function(/** @type {Date} */date) {
    __emval_run_destructors(date);
  },
}

mergeInto(LibraryManager.library,myLib);
```

Is it already possible? If not is it could be possible?

Alternatively, are there any compiler flag I could check to see if there is `Emval` object defined or I need to use `_emval_register` instead ?

**Comment by kripken** [comment]
You do have access to [`EMSCRIPTEN_VERSION`](https://github.com/emscripten-core/emscripten/blob/53948389b6e17f170113602f9bd95335ea59decb/src/settings_internal.js#L108) in js library files. But it is a string like `"3.1.36-git"` or `"3.1.36"`, so `<`, `>` comparisons would need to go through a helper function I guess.

**Reply by Thaina** [reply]
Thank you very much. Sadly after I have search for that macro, I have seen there is `EMSCRIPTEN_VERSION_MAJOR` exist in the python. But it seem it not passed into js compilation. Which would make that useful for version filtering

And if I understand correctly, I think we could not write helper function for compiler option like at the `#if`, didn't we?

**Reply by Thaina** [reply]
I have try modified emscripten in unity and this seem to work as I expected https://github.com/emscripten-core/emscripten/pull/19173

**Reply by sbc100** [reply]
Can you not just use `EMSCRIPTEN_VERSION` as @kripken suggests above?

**Reply by Thaina** [reply]
> Can you not just use `EMSCRIPTEN_VERSION` as @kripken suggests above?

@sbc100 Could you please enlighten me how to write a so called helper function to parse version string and compare it to number to have the same effect as this

```
#if EMSCRIPTEN_VERSION_MAJOR < 2 || (EMSCRIPTEN_VERSION_MAJOR == 2 && (EMSCRIPTEN_VERSION_MINOR < 0 || (EMSCRIPTEN_VERSION_MINOR == 0 && EMSCRIPTEN_VERSION_TINY < 31)))
  __emval_register
#else
  Emval.toHandle
#endif
```

in essence, I want to compare `EMSCRIPTEN_VERSION < 2.0.31` in the compiler macro, not in the js code. I don't see how could I even parse the version from string to numeric value in here

**Reply by sbc100** [reply]
I think you can write a new helper function to put it inline into your library file.   Something like this one maybe: 

https://github.com/emscripten-core/emscripten/blob/309581e433f114e47ce61d619fbdc00136e9de90/src/library_html5_webgpu.js#L1-L3

Actually I think in order for the function to be available to the preprocessor you might need to declare it in a previous library file.  So you would need to do something like `--js-library emscripten_version_macros.js --js-library my_js_lib.js`.

Maybe not the most elegant solution.. but I think it should work.

**Reply by Thaina** [reply]
And that's push another more complication to the compiler flags, while the numeric version already exist in python and the usefulness of the version is to be used in the preprocessor as version independent code

The numeric version would come in handy if there would be any breaking change feature in the future. I have see no reason to not add it

**Reply by sbc100** [reply]
I was think you just do something like this:

```
{{{
function splitVersion(v) {
  return v.split('.').slice(0, 3);  // something like this
}
function checkEmccVersion(v) {
  return splitVersion(EMSCRIPTEN_VERSION) >= splitVersion(v);
}
}}}
```

**Reply by sbc100** [reply]
Then in your JS libraries you can do: 

```
#if checkEmccVersion("2.3.4")
..
#else
..
#endif
```

The other upside of this approach is that it will work with older version of emscripten..  If we add new settings those can't be used until they are released and will never work in older versions.

**Reply by Thaina** [reply]
> I was think you just do something like this:
> 
> ```
> {{{
> function splitVersion(v) {
>   return v.split('.').slice(0, 3);  // something like this
> }
> function checkEmccVersion(v) {
>   return splitVersion(EMSCRIPTEN_VERSION) >= splitVersion(v);
> }
> }}}
> ```

Is it require any compiler flags or anything to make it work though, that's the main problem. I wish I could paste this code and it just work unlike `EMSCRIPTEN_KEEPALIVE`

**Reply by sbc100** [reply]
But this issue related to JS library file, no?

If you want to write JS code within C++ source files when you have the EM_JS and EM_ASM options, and you have the existing `_EMSCRIPTEN_major__`, `__EMSCRIPTEN_minor__` and `__EMSCRIPTEN_tiny__` macros (which are mentioned in the title of this issue).

**Reply by Thaina** [reply]
I am confused with these sentence

> declare it in a previous library file. So you would need to do something like --js-library emscripten_version_macros.js --js-library my_js_lib.js.

I try to confined the code needed in one js file and don't need to add anymore flag or compiler argument because it would be too complicate for unity build system. So what I am saying is do I need to do this or could I just paste your code in js file and it will work?

**Reply by sbc100** [reply]
I think the above example requires two different `--js-library` files.    If I understand you correctly you already have one `--js-library` file, is that right?   And you want to be able to do something like `#if checkEmccVersion("2.3.4")` within that file? 

As far as I know that only way to add a new preprocess helper is to put it in a separate `--js-library` file that gets specified before the one you want to use it in.

(or you could embed the JS inside you C/C++ sources with EM_ASM/EM_JS too).

**Reply by sbc100** [reply]
It sounds like you are saying that the unity build system allows your to add a `--js-libray` file, but not two of them?   Is that correct?

**Reply by Thaina** [reply]
> only way to add a new preprocess helper is to put it in a separate --js-library file that gets specified before the one you want to use it in

Right, that's why I ask

> Is it require any compiler flags or anything to make it work though

If it's yes, then it was not suitable for unity and any similar system that utilized emscripten in roundabout way. That's why I make this pull request so that, at least in the future, this problem will be no more

**Reply by Thaina** [reply]
> It sounds like you are saying that the unity build system allows your to add a `--js-libray` file, but not two of them? Is that correct?

Unity build system not prohibit appending any arguments to emscripten. But it's not obvious and require fine grain management on the internal build step. It also complicate and can be conflict from many system trying to manage same project

Not to mention unity also have their own internal flag and also internal step I don't know much about (which also seem to changed in each version (and each version of unity sometimes also change the version of emscripten)). The build process also taken place in it's own directory and referencing any file on compiler flag is confusing

That's why it's unwieldy to try messing up the emscripten compiler flag in unity. And so everything should be contained in the script file as much as possible

**Reply by Thaina** [reply]
@sbc100 OK after experimenting a bit more. I found out that we could actually just write string function directly and so this request is not really needed

```C
#if EMSCRIPTEN_VERSION.split('.').slice(0,4).map(function(n) { return parseInt(n) }) > [2,0,31]
		console.log("{{{EMSCRIPTEN_VERSION}}}" + " > 2.0.31");
#elif EMSCRIPTEN_VERSION.split('.').slice(0,4).map(function(n) { return parseInt(n) }) < [2,0,19]
		console.log("{{{EMSCRIPTEN_VERSION}}}" + " < 2.0.19");
#else
		console.log("{{{EMSCRIPTEN_VERSION}}}" + " >= 2.0.19");
#endif
```

Thank you very much for giving me all these explanations

**Reply by sbc100** [reply]
You might also want look into writing your JS code directly in the C/C++ source using EM_JS

**Reply by juj** [reply]
Thanks for this neat snippet @Thaina . I wonder where the `.slice(0,4)` part comes from?

I ended up using

```
#if EMSCRIPTEN_VERSION.split('.').map(parseInt) < [3,1,35]
```
in my [wasm_webgpu](https://github.com/juj/wasm_webgpu/commit/7a5ed94607559116f5e592ed2d83b84f7a2d53d6#diff-31deb58826d8c50a2e8b817f199ae92287a40dcd601f7ddde83d9f018ae927b4R1-R5) bindings library.

**Reply by Thaina** [reply]
> Thanks for this neat snippet @Thaina . I wonder where the `.slice(0,4)` part comes from?
> 
> I ended up using
> 
> ```
> #if EMSCRIPTEN_VERSION.split('.').map(parseInt) < [3,1,35]
> ```
> 
> in my [wasm_webgpu](https://github.com/juj/wasm_webgpu/commit/7a5ed94607559116f5e592ed2d83b84f7a2d53d6#diff-31deb58826d8c50a2e8b817f199ae92287a40dcd601f7ddde83d9f018ae927b4R1-R5) bindings library.

`slice(0,4)` means I would pick only 4 item of the split that most likely be a version number to parse as integer

**Reply by juj** [reply]
Big gotcha I found out today: `#if EMSCRIPTEN_VERSION.split('.').map(parseInt) < [3,1,35]` as I presented above does not actually work at all!

Instead, one must use

```
#if EMSCRIPTEN_VERSION.split('.').map(parseFloat) < [3,1,35]
```

Details of why `parseInt` does not work, but `parseFloat` does, can be found at https://stackoverflow.com/questions/262427/why-does-parseint-yield-nan-with-arraymap

**Reply by juj** [reply]
Actually, now that I look into this, the whole .map(parseInt/Float) thing doesn't work at all. When I read the above code first, I assumed that arrays had a lexicographic compare, but that is not the case.

Rather the arrays get converted back to strings, so this achieves nothing. https://stackoverflow.com/questions/8328908/javascript-surprising-array-comparison
